### DISCUSSION TITLE
Something very weird is going on with last day to capture a castle in the OG and it is better but weird in fheroes2 too

### DISCUSSION THREAD

**Discussion by zenseii** [ihhub/fheroes2#8305] [General]
After discussing in https://github.com/ihhub/fheroes2/pull/8301

Consider the following scenario:
5 heroes without a castle day 1.
2 players get eliminated.
The remaining 3 wait.

On the start of day 7 is when weird things start to happen in the OG: 
In the OG the first player, aka Blue, is told that they have 2 days left, meanwhile the other players are told that this is their last day. And this is actually the case - when they end their turns they lose, and when blue starts what is their final day, aka Day 1, Week 2, they are told that they have won. We can all agree that this is complete non-sense, and luckily in fheroes2 this is not the case currently.

Relevant save from OG (Go to multiplayer -> hotseat -> 3 players -> original map to find it)
[3PlayersCaptureDefeatOG.zip](https://github.com/ihhub/fheroes2/files/13986899/3PlayersCaptureDefeatOG.zip)

However, what I actually wanted to test before finding this bug, was what would happen, if we assume no bugs that the original game has which is the case for fheroes2, then if blue player ends their turn on the final day to capture a town, aka Day 1, Week 2, and then the other players get eliminated on the same day, for example if the 3rd player defeats the 4th player, and then the 3rd player dismisses their last hero removing all other players than Blue, or if all of the other players dismiss their heroes, or if they all attack Blue and all of them lose.

What I found out is that strangely Blue loses apparently. Even if in reality Blue is still in the game when all of these things happen, you always get the message that Blue player has been vanquished.

In my opinion Blue should be the winner, because all of this happened during the final day and Blue was still alive when the day ended in the last alive player's turn.

I would consider this a bug, but I'm opening this for discussion.

Here is the test save for fheroes2:
[DaysToCaptureStrangeFH2.zip](https://github.com/ihhub/fheroes2/files/13987078/DaysToCaptureStrangeFH2.zip)
Let blue end their turn on the final day and then somehow eliminate the other players during their turns.

Here is the map file, look for Multiplayer, Original, 5 players, small. (I originally made it to test surrender costs)
[SurrndrC.zip](https://github.com/ihhub/fheroes2/files/13987107/SurrndrC.zip)

Related discussions: https://github.com/ihhub/fheroes2/issues/7092 and https://github.com/ihhub/fheroes2/issues/2625

**Comment by zenseii** [comment]
@oleg-derevenetz, you're the expert on this topic. What are your thoughts about how the logic I described here works in fheroes2 currently?

**Reply by oleg-derevenetz** [reply]
Currently in fheroes2, win/loss conditions are checked for the currently active kingdom, because the relevant messages are formulated in such a way that they relate to the current active kingdom/player (e.g. "You captured some town. You are victorious." or "The enemy is beaten. Your side has triumphed!"). And when the blue kingdom becomes active in its turn, it has already lost, as the next day has come. For example, in this video:

https://github.com/ihhub/fheroes2/assets/32623900/cd06bc98-3290-46ca-b856-2d9d02c595fb

first, the game shows whose turn it is now, and then the winning video is shown (or the winning message in case of other victory conditions), that is, it becomes obvious who won.

**Reply by zenseii** [reply]
Thanks for the input, @oleg-derevenetz.

>because the relevant messages are formulated in such a way that they relate to the current active kingdom/player

This is true, though they do also show the color's crest/coat of arms if I'm not mistaken, so this could make it less strange that it is referring to "you". However, even if it wouldn't be strange to have such a dialog pop up in Hot Seat matches, it would be weird, but not ambiguous, in an online multiplayer match for Purple player to see a message talking about "you winning" and they are the only person in the room, even if there's a colored coat of arms in the dialog.

Consider this case in the OG where it's Yellow's turn and they lose to Purple. The game doesn't end the turn, because it doesn't display the "Purple player's turn", and Purple is given the win before their turn starts:

https://github.com/ihhub/fheroes2/assets/12501091/a92c52e1-0a95-4d25-8382-ad0ec02ffaa8

Whereas in fheroes2 when Yellow dies in their own turn, the turn is automatically ended after the battle and Purple gets their turn and is then notified about winning:

https://github.com/ihhub/fheroes2/assets/12501091/e11d3e08-5c12-4e63-b5e8-45495e8bc1bc


I guess my main concern is that by correcting how the OG works we also introduced a scenario in which there is no winner, aka a tie, which I don't think has ever existed in the Heroes series?

EDIT: Maybe there could be cases for scenarios where you're supposed to accrue a certain amount of gold to win, it would be possible for two players to reach this amount on the same day. My hunch is that the player that comes first in the player order gets the win. This is just one of the strange effects of a turn-based game probably.

**Reply by oleg-derevenetz** [reply]
> I guess my main concern is that by correcting how the OG works we also introduced a scenario in which there is no winner, aka a tie, which I don't think has ever existed in the Heroes series?

How is that? It has always existed, consider:

https://github.com/ihhub/fheroes2/assets/32623900/d837a4f2-32be-43ea-9b04-bd4c6f46e819

There is no winner in this multiplayer scenario. This may happen in any scenario with non-default victory conditions (e.g. "capture a castle", "eliminate a hero", "find an artifact", etc), and in fheroes2 this may happen also with default win/loss conditions in certain circumstances. I don't see any great tragedy in this.

> Maybe there could be cases for scenarios where you're supposed to accrue a certain amount of gold to win, it would be possible for two players to reach this amount on the same day. My hunch is that the player that comes first in the player order gets the win.

Yes, in turn-based strategies, the first player always has some advantage, that's how they work.

**Reply by zenseii** [reply]
> I don't see any great tragedy in this.

In fact I think it would be a great comedy to experience 😂

Imagine that on the dawn of the final day, after player 1's turn, players 3 and 4 decide to attack blue, but player 1 plays like a champion and defeats both of the enemy heroes against all odds. However, just as they get ready to celebrate their great achievement they get presented with the lose screen... and they flip the table and karate chop the keyboard in half 😂 Basically it wouldn't have made any difference whether they won the battles or just retreated in them, or just dismissed their hero at the end of their turn. Their continued existence beyond their last turn that day was meaningless for them that day.

https://github.com/ihhub/fheroes2/assets/12501091/5fc46f33-701d-4910-907c-6e7adf4adc0e


It's even better if we consider that the default win/loss conditions state that you win by "Defeating all enemy heroes" which they clearly did but too bad.
<img width="377" alt="image" src="https://github.com/ihhub/fheroes2/assets/12501091/61667a81-faab-42d5-b99f-5673967e7eb7">

**Reply by oleg-derevenetz** [reply]
> However, just as they get ready to celebrate their great achievement they get presented with the lose screen... and they flip the table and karate chop the keyboard in half 😂

Well, at the end of his last day, he should have received a warning that "your heroes abandon you, and you are banished from this land", so It shouldn't come as a big surprise to him.
